test/extensions/tracers/opencensus/tracer_test.cc

```diff
@@ -165,74 +166,86 @@ TEST(OpenCensusTracerTest, Span) {
+            {":method", "GET"},
+            {"x-request-id", "foo"},
+        };
+        for (const auto& kv : headers) {
+          request_headers.addCopy(LowerCaseString(kv.first), kv.second);
+        }
+
+        const std::string operation_name{"my_operation_2"};
+        SystemTime start_time;
+        Http::TestHeaderMapImpl injected_headers;
+        {
+          Tracing::SpanPtr span = driver->startSpan(config, request_headers, operation_name,
+                                                    start_time, {Tracing::Reason::Sampling, false});
+          span->injectContext(injected_headers);
+          span->finishSpan();
+        }
+
+        // Retrieve SpanData from the OpenCensus trace exporter.
+        std::vector<SpanData> spans = getSpanCatcher()->catchSpans();
+        ASSERT_EQ(1, spans.size());
+        const auto& sd = spans[0];
+        ENVOY_LOG_MISC(debug, "{}", sd.DebugString());
+
+        // Check contents.
+        EXPECT_TRUE(sd.has_remote_parent());
+        EXPECT_EQ("6162636465666768", sd.parent_span_id().ToHex());
+        EXPECT_EQ("404142434445464748494a4b4c4d4e4f", sd.context().trace_id().ToHex());
+        EXPECT_TRUE(sd.context().trace_options().IsSampled())
+            << "parent was sampled, child should be also";
+
+        // Check injectContext.
+        // The SpanID is unpredictable so re-serialize context to check it.
+        const auto& ctx = sd.context();
+        auto test_outgoing = [&injected_headers](const char* header, std::string expected_value) {
+          auto val = injected_headers.get(LowerCaseString(header));
+          ASSERT_NE(nullptr, val);
+          EXPECT_EQ(expected_value, val->value().getStringView());
+        };
+        test_outgoing("traceparent", ::opencensus::trace::propagation::ToTraceParentHeader(ctx));
+        {
+          std::string expected = ::opencensus::trace::propagation::ToGrpcTraceBinHeader(ctx);
+          expected = Base64::encode(expected.data(), expected.size(), /*add_padding=*/false);
+          test_outgoing("grpc-trace-bin", expected);
+        }
+        test_outgoing("x-cloud-trace-context",
+                      ::opencensus::trace::propagation::ToCloudTraceContextHeader(ctx));
+        test_outgoing("x-b3-traceid", "404142434445464748494a4b4c4d4e4f");
+        test_outgoing("x-b3-spanid", ::opencensus::trace::propagation::ToB3SpanIdHeader(ctx));
+        test_outgoing("x-b3-sampled", "1");
+      };
+
+  test_incoming({{"traceparent", "00-404142434445464748494a4b4c4d4e4f-6162636465666768-01"}});
+  test_incoming({{"grpc-trace-bin", "AABAQUJDREVGR0hJSktMTU5PAWFiY2RlZmdoAgE"}});
+  test_incoming(
+      {{"x-cloud-trace-context", "404142434445464748494a4b4c4d4e4f/7017280452245743464;o=1"}});
+  test_incoming({{"x-b3-traceid", "404142434445464748494a4b4c4d4e4f"},
+                 {"x-b3-spanid", "6162636465666768"},
+                 {"x-b3-sampled", "1"}});
+  test_incoming({{"x-b3-traceid", "404142434445464748494a4b4c4d4e4f"},
```

Comment:
Is it possible to move these tests away from this Javascript-like nested callbacks to independent TEST cases with some helper functions? I can see there is some common stuff here, but I think it's more readable in normative C++. I think each trace header type can have its own TEST this way. Thanks.

Reply:
Done.